apm-agent-plugins/apm-log-shader-plugin/apm-log-shader-plugin-common/src/main/java/co/elastic/apm/agent/log/shader/AbstractEcsReformattingHelper.java

```diff
@@ -0,0 +1,447 @@
+ *          if OVERRIDE:
+ *              if there is no entry in the originalAppender2originalFormatter map, then (thread safely):
+ *                  create an ECS-formatter
+ *                  add an entry to originalAppender2originalFormatter map
+ *                  add an entry to originalAppender2ecsFormatter map
+ *                  replace original formatter with ECS-formatter in current appender
+ *          else
+ *              if the current appender uses an ECS-formatter (thread safely):
+ *                  restore original formatter in current appender
+ *                  remove entry from originalAppender2originalFormatter map
+ *                  remove entry from originalAppender2ecsFormatter map
+ *              if SHADE or REPLACE:
+ *                  if there is no entry in the originalAppender2ecsAppender mpa, then (thread safely):
+ *                      create an ECS-appender
+ *                      add entry to originalAppender2ecsAppender map
+ *                  if REPLACE and there is a valid ECS-appender:
+ *                      skip append() execution on original appender
+ *      on getLayout() exit:
+ *          * only relevant to log4j2, see {@code co.elastic.apm.agent.log4j2.Log4j2AppenderGetLayoutAdvice} for details
+ *          if OVERRIDE:
+ *              return the ECS-formatter instead of the original formatter
+ *      {@link #onAppendExit(Object) on append() exit}:
+ *          if SHADE or REPLACE:
+ *              invoke append() on ECS-appender
+ *          clear log_ecs_reformatting config from thread local
+ * </pre>
+ * <br>
+ *
+ * @param <A> logging-framework-specific Appender type
+ * @param <F> logging-framework-specific formatter type ({@code Layout} in Log4j, {@code Encoder} in Logback)
+ */
+@GlobalState
+public abstract class AbstractEcsReformattingHelper<A, F> {
+
+    // Escape shading
+    private static final String ECS_LOGGING_PACKAGE_NAME = "co!elastic!logging".replace('!', '.');
+
+    // We can use regular shaded logging here as this class is loaded from the agent CL
+    private static final Logger logger = LoggerFactory.getLogger(AbstractEcsReformattingHelper.class);
+    public static final String ECS_SHADE_APPENDER_NAME = "EcsShadeAppender";
+
+    // Used to cache the fact that ECS-formatter or ECS-appender are not created for a given appender
+    private static final Object NULL_APPENDER = new Object();
+    private static final Object NULL_FORMATTER = new Object();
+
+    private static final CallDepth callDepth = CallDepth.get(AbstractEcsReformattingHelper.class);
+
+    /**
+     * A mapping between original appender and the corresponding ECS-appender.
+     * Used when {@link LoggingConfiguration#logEcsReformatting log_ecs_reformatting} is set to
+     * {@link LogEcsReformatting#SHADE SHADE} or {@link LogEcsReformatting#REPLACE REPLACE}.
+     */
+    private static final WeakConcurrentMap<Object, Object> originalAppender2ecsAppender = WeakMapSupplier.createMap();
```

Comment:
A habit 🙂 
This guarantees there are no duplication, no matter where and how many helpers are created. Conceptually, everything here could be static, but with the maps this may actually have an effect if not.
Any reason to make them non-static? Are you worried from higher contention or something?

Reply:
If they weren't static, we could use generics `WeakConcurrentMap<A, A>`/`WeakConcurrentMap<A, R>` which could make things a bit more readable. As we already have a static singleton instance, I was just a bit confused why the maps need to be shared across the different implementations.
That's not a blocker though.

Reply:
It's a good point, but since I use dummy formatter and appender to optimize lookups that can't or shouldn't result with real implementations, switching to generic maps means adding the complication of creating these dummies in the helper subclasses. I like the fact that it is hidden from the sub-helpers, but I can try this out